std/range/package.d

```diff
@@ -4688,7 +4688,13 @@ if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
         && isIntegral!S)
 in
 {
-    assert(!(step == 0 && begin != end));
+    import std.conv : unsigned;
+
+    assert(step != 0 || begin == end);
+    if (step > 0)
+        assert(unsigned((end - begin) / step) <= size_t.max);
+    else if (step < 0)
+        assert(unsigned((end - begin) / -step) <= size_t.max);
```

Comment:
I think we should also have an assert that the begin and end are properly related. That is, if step < 0, having begin < end is OK, it shouldn't fail (because it's immediately empty).